src/bos/common/clients/bos/base.py

```diff
@@ -0,0 +1,121 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+from abc import ABC
+import logging
+
+from bos.common.clients.endpoints import BaseEndpoint
+from bos.common.tenant_utils import get_new_tenant_header
+from bos.common.utils import PROTOCOL
+
+LOGGER = logging.getLogger(__name__)
+
+API_VERSION = 'v2'
+SERVICE_NAME = 'cray-bos'
+BASE_BOS_ENDPOINT = f"{PROTOCOL}://{SERVICE_NAME}/{API_VERSION}"
+
+
+class BaseBosEndpoint(BaseEndpoint, ABC):
+    """
+    This base class provides generic access to the BOS API.
+    The individual endpoint needs to be overridden for a specific endpoint.
+    """
+    BASE_ENDPOINT = BASE_BOS_ENDPOINT
+
+
+class BaseBosNonTenantAwareEndpoint(BaseBosEndpoint, ABC):
+    """
+    This base class provides generic access to the BOS API for non-tenant-aware endpoints
+    The individual endpoint needs to be overridden for a specific endpoint.
+    """
+
+    def get_item(self, item_id):
+        """Get information for a single BOS item"""
+        return self.get(uri=item_id)
+
+    def get_items(self, **kwargs):
+        """Get information for all BOS items"""
+        return self.get(params=kwargs)
+
+    def update_item(self, item_id, data):
+        """Update information for a single BOS item"""
+        return self.patch(uri=item_id, json=data)
+
+    def update_items(self, data):
+        """Update information for multiple BOS items"""
+        return self.patch(json=data)
+
+    def put_items(self, data):
+        """Put information for multiple BOS Items"""
+        return self.put(json=data)
+
+    def delete_items(self, **kwargs):
+        """Delete information for multiple BOS items"""
+        return self.delete(params=kwargs)
+
+
+class BaseBosTenantAwareEndpoint(BaseBosEndpoint, ABC):
+    """
+    This base class provides generic access to the BOS API for tenant aware endpoints.
+    The individual endpoint needs to be overridden for a specific endpoint.
+    """
+
+    def get_item(self, item_id, tenant):
+        """Get information for a single BOS item"""
+        return self.get(uri=item_id, headers=get_new_tenant_header(tenant))
+
+    def get_items(self, **kwargs):
+        """Get information for all BOS items"""
+        headers = None
+        if "tenant" in kwargs:
+            tenant = kwargs.pop("tenant")
+            headers = get_new_tenant_header(tenant)
+        return self.get(params=kwargs, headers=headers)
+
+    def update_item(self, item_id, tenant, data):
+        """Update information for a single BOS item"""
+        return self.patch(uri=item_id,
```

Comment:
position parameter uri is actually an uri or item_id?

Reply:
I am not sure what you are asking.
For the update_item function, it's an item ID.
That function translates from the high level objective ("update the item with this ID belonging to this tenant, with the following data") to the API level request ("Perform a patch request to this URI ...").
It happens that the URI in this case is the ID.

Reply:
ok parameter name and supplied value are causing confusion. I suppose the URI is formed using the item ID. may be the parameter shoudl be renamed to "item_id"